### PR TITLE
[Linux consumption] Reinitialize MSI sidecar on container restarts

### PR DESCRIPTION
**Pull request by balag0**
Backport https://github.com/Azure/azure-functions-host/commit/82638e5f08700f9620f6141dc85ab142c848d505

### Pull request checklist

* [X ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ X] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ X] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ X] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs [hunk 1]
```diff
@@ -55,6 +55,15 @@ private async Task ApplyStartContextIfPresent()
 
                 bool success = _instanceManager.StartAssignment(assignmentContext);
                 _logger.LogInformation($"StartAssignment invoked (Success={success})");
+
+                if (assignmentContext.MSIContext != null)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs]: Why is this logic different for V2? The V3 PR doesn't null check this I see. Ah - I see, you've sent a new V3 PR that adds a null check

File: src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs [hunk 2]
```diff
@@ -55,6 +55,15 @@ private async Task ApplyStartContextIfPresent()
...
+                {
+                    await _instanceManager.SpecializeMSISidecar(assignmentContext);
+                }
+                else
+                {
+                    _logger.LogInformation("Skipping re-specialization of MSI sidecar.");
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs]: As in the original PR, can't we just leave any required logging in SpecializeMSISidecar? Or are you saying that you want a log statement if MSI info isn't present? Perhaps you should put this null check and log statement in SpecializeMSISidecar so we're sure to do it in all cases

File: src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs (hunk 1 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs]: Yes :)

File: src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs (hunk 2 above)
Comment (balag0) [reply src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs]: Yes, adding an additional log here lets us query by Source and possibly use it in a detector.
This is an exception scenario and not expected to be hit often.

Comment (mathewc) [reply src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerInitializationHostService.cs]: Then I'd say move the null check into SpecializeMSISidecar to centralize it. "Skipping specialization of MSI sidecar - MSI info is empty" or something like that

**Review by mathewc — DISMISSED** [review]

**Review by mathewc — APPROVED** [review]

**Pull request merged**